src/Advection/weno5.jl

```diff
@@ -0,0 +1,274 @@
+@inline βy₁(i, j, k, f) = @inbounds 13/12 * (f[i, j-2, k] - 2f[i, j-1, k] + f[i, j,   k])^2 + 1/4 * ( f[i, j-2, k]                 -  f[i,   j, k])^2
+@inline βy₂(i, j, k, f) = @inbounds 13/12 * (f[i, j-3, k] - 2f[i, j-2, k] + f[i, j-1, k])^2 + 1/4 * ( f[i, j-3, k] - 4f[i, j-2, k] + 3f[i, j-1, k])^2
+
+@inline βz₀(i, j, k, f) = @inbounds 13/12 * (f[i, j, k-1] - 2f[i, j,   k] + f[i, j, k+1])^2 + 1/4 * (3f[i, j, k-1] - 4f[i, j,   k] +  f[i, j, k+1])^2
+@inline βz₁(i, j, k, f) = @inbounds 13/12 * (f[i, j, k-2] - 2f[i, j, k-1] + f[i, j,   k])^2 + 1/4 * ( f[i, j, k-2]                 -  f[i, j,   k])^2
+@inline βz₂(i, j, k, f) = @inbounds 13/12 * (f[i, j, k-3] - 2f[i, j, k-2] + f[i, j, k-1])^2 + 1/4 * ( f[i, j, k-3] - 4f[i, j, k-2] + 3f[i, j, k-1])^2
+
+#####
+##### WENO-5 optimal weights
+#####
+
+const C3₀ = 3/10
+const C3₁ = 3/5
+const C3₂ = 1/10
+
+#####
+##### WENO-5 raw weights
+#####
+
+const ε = 1e-6
+const ƞ = 2  # WENO exponent
+
+@inline αx₀(i, j, k, f) = C3₀ / (βx₀(i, j, k, f) + ε)^ƞ
+@inline αx₁(i, j, k, f) = C3₁ / (βx₁(i, j, k, f) + ε)^ƞ
+@inline αx₂(i, j, k, f) = C3₂ / (βx₂(i, j, k, f) + ε)^ƞ
+
+@inline αy₀(i, j, k, f) = C3₀ / (βy₀(i, j, k, f) + ε)^ƞ
+@inline αy₁(i, j, k, f) = C3₁ / (βy₁(i, j, k, f) + ε)^ƞ
+@inline αy₂(i, j, k, f) = C3₂ / (βy₂(i, j, k, f) + ε)^ƞ
+
+@inline αz₀(i, j, k, f) = C3₀ / (βz₀(i, j, k, f) + ε)^ƞ
+@inline αz₁(i, j, k, f) = C3₁ / (βz₁(i, j, k, f) + ε)^ƞ
+@inline αz₂(i, j, k, f) = C3₂ / (βz₂(i, j, k, f) + ε)^ƞ
+
+#####
+##### WENO-5 normalized weights
+#####
+
+@inline function weno5_weights_x(i, j, k, f)
+    α₀ = αx₀(i, j, k, f)
+    α₁ = αx₁(i, j, k, f)
+    α₂ = αx₂(i, j, k, f)
+    
+    Σα = α₀ + α₁ + α₂ 
+    w₀ = α₀ / Σα
+    w₁ = α₁ / Σα
+    w₂ = α₂ / Σα
+    
+    return w₀, w₁, w₂
+end
+
+@inline function weno5_weights_y(i, j, k, f)
+    α₀ = αy₀(i, j, k, f)
+    α₁ = αy₁(i, j, k, f)
+    α₂ = αy₂(i, j, k, f)
+
+    Σα = α₀ + α₁ + α₂
+    w₀ = α₀ / Σα
+    w₁ = α₁ / Σα
+    w₂ = α₂ / Σα
+    
+    return w₀, w₁, w₂
+end
+
+@inline function weno5_weights_z(i, j, k, f)
+    α₀ = αz₀(i, j, k, f)
+    α₁ = αz₁(i, j, k, f)
+    α₂ = αz₂(i, j, k, f)
+
+    Σα = α₀ + α₁ + α₂
+    w₀ = α₀ / Σα
+    w₁ = α₁ / Σα
+    w₂ = α₂ / Σα
+    
+    return w₀, w₁, w₂
+end
+
+#####
+##### WENO-5 flux reconstruction
+#####
+
+@inline function weno5_flux_x(i, j, k, f)
+    w₀, w₁, w₂ = weno5_weights_x(i, j, k, f)
+    return w₀ * px₀(i, j, k, f) + w₁ * px₁(i, j, k, f) + w₂ * px₂(i, j, k, f)
+end
+
+@inline function weno5_flux_y(i, j, k, f)
+    w₀, w₁, w₂ = weno5_weights_y(i, j, k, f)
+    return w₀ * py₀(i, j, k, f) + w₁ * py₁(i, j, k, f) + w₂ * py₂(i, j, k, f)
+end
+
+@inline function weno5_flux_z(i, j, k, f)
+    w₀, w₁, w₂ = weno5_weights_z(i, j, k, f)
+    return w₀ * pz₀(i, j, k, f) + w₁ * pz₁(i, j, k, f) + w₂ * pz₂(i, j, k, f)
+end
+
+#####
+##### Momentum advection fluxes
+#####
+
+@inline momentum_flux_uu_weno5(i, j, k, grid, u)    = ℑxᶜᵃᵃ(i, j, k, grid, Ax_ψᵃᵃᶠ, u) * weno5_flux_x(i, j, k, u)
```

Comment:
This comment is nitpicky for sure, but easy to address: I think in the other files we use the English-language convention "fourth_order_momentum_flux" for functions like this. For this function that'd mean calling it "weno5_momentum_flux". I'm fine to change the other functions to use a suffix style but its probably best to be consistent.

Reply:
We may want:

```julia
@inline momentum_flux_uu_weno5(i, j, k, grid, u)    = ℑxᶜᵃᵃ(i, j, k, grid, Ax_ψᵃᵃᶠ, u) * weno5_flux_x(i+1, j, k, u)
```

and similarly for all fluxes that use `weno5_flux_x`, but are supposed to be defined at cell _centers_ rather than cell _interfaces_ (as the tracer fluxes are).

When this flux is differenced by `div_uu`, a backwards difference will be used; something to the effect of

```julia
momentum_flux_uu_weno5(i, j, k, grid, u) - momentum_flux_uu_weno5(i-1, j, k, grid, u)
```

Reply:
Resolved since these functions don't exist anymore.